RIOT-OS/RIOT
# drivers: add driver for Synopsys DWC2 that is used as USB OTG peripheral on STM32 and ESP32x SoCs

### Contribution description

This PR adds a driver for the Synopsys DWC2 IP core. It is an alternative to PR #18624 which avoids a lot of code duplication. STM32 as well as ESP32x SoCs use the Synopsys DWC2 core. According to [tinyUSB documentation](https://github.com/hathach/tinyusb/blob/master/docs/reference/supported.rst), EFM32GG12 and GD32VF103 also integrate the Synopsys DWC2 core. 

Instead of copying the code from STM32's `periph/usbdev_otg.c` and modifying it very slightly only for ESP32x SoCs (as done in PR #18624), the code is now moved to a common location and extended by only a very small number of MCU-Specific changes. The Namespace is changed from `stm32` to `dwc2`. STM32 and ESP32x SoCs use the driver as `periph_usbdev` driver.

### Testing procedure

The following tests have to work for an ESP32 node as well as a STM32 node to check whether it is still working for STM32.
1. Flash `tests/usbus_hid`
    ```
    BOARD=esp32s3-devkit make -C tests/usbus_hid flash term
    ```
    ```
    BOARD=nucleo-f767zi make -C tests/usbus_hid flash term
    ```
    Use the `dmesg` command and check the USB HID devices was found
    ```python
    [1230033.363705] usb 1-2.4: USB disconnect, device number 91
    [1230033.625919] usb 1-2.4: new full-speed USB device number 92 using xhci_hcd
    [1230033.732624] usb 1-2.4: New USB device found, idVendor=1209, idProduct=7d01, bcdDevice= 1.00
    [1230033.732627] usb 1-2.4: New USB device strings: Mfr=3, Product=2, SerialNumber=4
    [1230033.732629] usb 1-2.4: Product: esp32s3-devkit
    [1230033.732630] usb 1-2.4: Manufacturer: RIOT-os.org
    [1230033.732631] usb 1-2.4: SerialNumber: A5894EECB420ECF1
    [1230033.739837] hid-generic 0003:1209:7D01.00D9: hiddev2,hidraw5: USB HID v1.10 Device [RIOT-os.org esp32s3-devkit] on usb-0000:00:14.0-2.4/input0
    ```
    and that there is a new `/dev/hidrawX` device to which you can write as `root` characters that are echoed in console window, for example:
    ```
    # echo -n "Hello" > /dev/hidraw5
    ```
    ```
    main(): This is RIOT! (Version: 2022.10-devel-619-gca66ff-cpu/esp32/periph_usbdev)
    RIOT USB HID echo test
    Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here
    USB_HID rx_cb: Test argument 
    Msg received via USB HID: 
    Hello
    ```
2. Flash `tests/usbus_ecm`
    ```
    BOARD=esp32s3-devkit make -C tests/usbus_hid flash term
    ```
    ```
    BOARD=nucleo-f767zi make -C tests/usbus_hid flash term
    ```
   On the host computer, using tools such as `ethtool` must show the USB CDC ECM interface as link detected:
   ```shell
   # ethtool enp0s20u9u4
   Settings for enp0s20u9u4:
           Current message level: 0x00000007 (7)
                                  drv probe link
           Link detected: yes
   ```
   You should be able to ping the ESP32x node from the host and vice versa using the IPv6 address of the USB CDC ECM interface. 

### Issues/PRs references
